src/abi/substrate.rs

```diff
@@ -1,437 +1,465 @@
+    TypeDefVariant, Variant,
+};
 use semver::Version;
-use serde::{Deserialize, Serialize};
-use serde_json::{Map, Value};
 use solang_parser::pt;
-use std::convert::TryInto;
 
-#[derive(Deserialize, Serialize)]
-pub struct Abi {
-    storage: Storage,
-    types: Vec<Type>,
-    pub spec: Spec,
-}
+use crate::sema::{
+    ast::{self, ArrayLength, EventDecl, Function},
+    tags::render,
+};
 
-impl Abi {
-    pub fn get_function(&self, name: &str) -> Option<&Message> {
-        self.spec.messages.iter().find(|m| name == m.name)
-    }
+fn path_from_str(s: impl Into<String>) -> Path<PortableForm> {
+    serde_json::from_value::<Path<PortableForm>>(serde_json::json!(vec![s.into()])).unwrap()
 }
 
-#[derive(Deserialize, Serialize, PartialEq, Eq)]
-pub struct ArrayDef {
-    array: Array,
-}
-
-#[derive(Deserialize, Serialize, PartialEq, Eq)]
-pub struct Array {
-    len: usize,
-    #[serde(rename = "type")]
-    ty: usize,
-}
-
-#[derive(Deserialize, Serialize, PartialEq, Eq)]
-pub struct SequenceDef {
-    sequence: Sequence,
-}
-
-#[derive(Deserialize, Serialize, PartialEq, Eq)]
-pub struct Sequence {
-    #[serde(rename = "type")]
-    ty: usize,
-}
-
-#[derive(Deserialize, Serialize, PartialEq, Eq)]
-#[serde(untagged)]
-enum Type {
-    Builtin { def: PrimitiveDef },
-    BuiltinArray { def: ArrayDef },
-    BuiltinSequence { def: SequenceDef },
-    Struct { path: Vec<String>, def: Composite },
-    Enum { path: Vec<String>, def: EnumDef },
-}
-
-#[derive(Deserialize, Serialize, PartialEq, Eq)]
-struct BuiltinType {
-    id: String,
-    def: String,
-}
-
-#[derive(Deserialize, Serialize, PartialEq, Eq)]
-struct EnumVariant {
-    name: String,
-    discriminant: usize,
-}
-
-#[derive(Deserialize, Serialize, PartialEq, Eq)]
-struct EnumDef {
-    variant: Enum,
-}
-
-#[derive(Deserialize, Serialize, PartialEq, Eq)]
-struct Enum {
-    variants: Vec<EnumVariant>,
+fn primitive_to_ty(ty: &ast::Type, registry: &mut PortableRegistry) -> Type<PortableForm> {
+    match ty {
+        ast::Type::Int(_) | ast::Type::Uint(_) => int_to_ty(ty, registry),
+        ast::Type::Bool => Type::new(
+            path_from_str("bool"),
+            vec![],
+            TypeDef::Primitive(TypeDefPrimitive::Bool),
+            Default::default(),
+        ),
+        ast::Type::String => Type::new(
+            path_from_str("string"),
+            vec![],
+            TypeDef::Primitive(TypeDefPrimitive::Str),
+            Default::default(),
+        ),
+        _ => unreachable!("non primitive types"),
+    }
 }
 
-#[derive(Deserialize, Serialize, PartialEq, Eq)]
-struct Composite {
-    composite: StructFields,
-}
+fn int_to_ty(ty: &ast::Type, registry: &mut PortableRegistry) -> Type<PortableForm> {
+    let scalety = match ty {
+        // Substrate doesn't like primitive types which aren't a power of 2
+        // The abi encoder/decoder fixes this automatically
+        ast::Type::Uint(n) => format!("u{}", n.next_power_of_two()),
```

Comment:
Nit: rather than using `format!()` to create a string and then matching the string, you could do:

```rust
ast::Type::Uint(n) => match n.next_power_of_two()) {
        8 => TypeDefPrimitive::U8,
        16 => TypeDefPrimitive::U16,
        32 => TypeDefPrimitive::U32,
        64 => TypeDefPrimitive::U64,
        128 => TypeDefPrimitive::U128,
        256 => TypeDefPrimitive::U256,
},
```

Reply:
Semi adjusted, the sign part is still needed to generate the path name.

Reply:
I don't know why that is needed, you could write it like to so:

```rust
    let def = match ty {
        ast::Type::Uint(bits) => match bits.next_power_of_two() {
            8 => TypeDefPrimitive::U8,
            16 => TypeDefPrimitive::U16,
            32 => TypeDefPrimitive::U32,
            64 => TypeDefPrimitive::U64,
            128 => TypeDefPrimitive::U128,
            256 => TypeDefPrimitive::U256,
            _ => unreachable!(),
        },
        ast::Type::Int(bits) => match bits.next_power_of_two()) {
            8 => TypeDefPrimitive::I8,
            16 => TypeDefPrimitive::I16,
            32 => TypeDefPrimitive::I32,
            64 => TypeDefPrimitive::I64,
            128 => TypeDefPrimitive::I128,
            256 => TypeDefPrimitive::I256,
            _ => unreachable!(),
        },`
        _ => unreachable!(),
    };
```

Reply:
Ah I now see did you need the string later on

```rust
    let path = serde_json::from_value::<Path<PortableForm>>(serde_json::json!(vec![format!(
        "{signed}{scalety}"
    )]))
    .unwrap();
```
Nevermind, maybe the original was nicer. Up to you. Sorry for the bad review comments

Reply:
It's okay, I think it's not much different.